pallets/auctions/src/lib.rs

```diff
@@ -0,0 +1,726 @@
+            Self::deposit_event(Event::AuctionCancelled {
+                nft_id,
+                marketplace_id: auction.marketplace_id,
+                creator: who,
+            });
+
+            Ok(().into())
+        }
+
+        #[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+        #[transactional]
+        pub fn add_bid(
+            origin: OriginFor<T>,
+            nft_id: NFTId,
+            amount: BalanceOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let current_block = frame_system::Pallet::<T>::block_number();
+
+            // fetch the data of given nftId
+            let nft_data = T::NFTHandler::get_nft(nft_id).ok_or(Error::<T>::NFTIdInvalid)?;
+
+            // fetch data of auction that lists NFT
+            let auction = Auctions::<T>::get(nft_id).ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+            // ensure the nft is in listed for sale state
+            ensure!(
+                nft_data.listed_for_sale == true,
+                Error::<T>::NFTNotListedForSale
+            );
+
+            // ensure the caller is not the owner of NFT
+            ensure!(
+                nft_data.owner != who.clone(),
+                Error::<T>::OwnerCannotCreateBid
+            );
+
+            // ensure the auction period has commenced
+            ensure!(
+                auction.start_block < current_block,
+                Error::<T>::AuctionNotStarted
+            );
+
+            // ensure the auction period has not ended
+            ensure!(auction.end_block >= current_block, Error::<T>::AuctionEnded);
+
+            // ensure the bid amount is greater than start price
+            ensure!(auction.start_price < amount, Error::<T>::InvalidBidAmount);
+
+            // ensure the bid is larger than the current highest bid
+            if let Some(current_highest_bid) = auction.bidders.get_highest_bid() {
+                ensure!(amount > current_highest_bid.1, Error::<T>::InvalidBidAmount);
+            }
+
+            // reject if user already has a bid
+            ensure!(
+                auction.bidders.find_bid(who.clone()).is_none(),
+                Error::<T>::UserBidAlreadyExists
+            );
+
+            // transfer funds from caller
+            T::Currency::transfer(&who, &Self::account_id(), amount, KeepAlive)?;
+
+            // add bid to storage
+            Auctions::<T>::try_mutate(nft_id, |maybe_auction| -> DispatchResult {
+                // should not panic when unwrap since already checked above
+                let auction = maybe_auction
+                    .as_mut()
+                    .ok_or(Error::<T>::AuctionDoesNotExist)?;
+
+                // extend auction by grace period if in ending period
+                if auction.end_block.checked_sub(&current_block)
+                    <= Some(T::AuctionEndingPeriod::get())
+                {
+                    auction.end_block = auction
```

Comment:
This is flawed. If an auction lasts for 1 week, and we get an bid in the last 10 minutes we increase the duration of the auction up to 10 minutes. So if the bid was given 2 minutes before closing, we increase the duration for 8 minutes (so that the total is again 10 minutes).